### PR TITLE
[#169] Add UserRole model

### PR DESCRIPTION
**Pull request by kmarszalek**
- change AppRole to enum
- add UserRole Entity: role, user, group
- removed old code
- add @transactional(readonly = true) on db repositories
- change AppUser to fields inside UserDetails

### CODE REVIEW

File: s4e-backend/src/main/java/pl/cyfronet/s4e/service/GroupService.java [hunk 1]
```diff
@@ -110,4 +174,41 @@ public void update(Group group) throws GroupUpdateException {
             throw new GroupUpdateException("Cannot update Group", e);
         }
     }
+
+    @Transactional(rollbackFor = GroupUpdateException.class)
+    public void updateFromRequest(UpdateGroupRequest request, String institutionSlug, String groupSlug) throws GroupUpdateException, NotFoundException {
+        Set<String> membersEmailsFromDB = groupRepository.findAllMembersEmails(institutionSlug, groupSlug, AppRole.GROUP_MEMBER);
+        Set<String> difference = membersEmailsFromDB.stream().collect(Collectors.toSet());
```
Comment (jswk) [inline s4e-backend/src/main/java/pl/cyfronet/s4e/service/GroupService.java]: Why is the `difference` variable not initialized directly from the repo?
I'd call the `difference` maybe `membersToRemove`, `difference` is unclear.

File: s4e-backend/src/main/java/pl/cyfronet/s4e/service/GroupService.java [hunk 2]
```diff
@@ -49,41 +90,52 @@ public Group save(Group group) throws GroupCreationException {
...
 
-    private void removeFromGroups(Set<Group> groupsFromRequest, AppUser user, Set<Group> userGroups) {
-        for (Group group: userGroups) {
-            if (!groupsFromRequest.contains(group)) {
-                group.removeMember(user);
-                user.getGroups().remove(group);
+    @Transactional
+    public void removeFromGroups(Set<Group> groupsFromRequest, AppUser user, Set<Group> userGroups) {
+        for (Group group : userGroups) {
+            if (groupsFromRequest.isEmpty() || !groupsFromRequest.contains(group)) {
```
Comment (jswk) [inline s4e-backend/src/main/java/pl/cyfronet/s4e/service/GroupService.java]: Why the first condition? It holds whenever the second one does.

File: s4e-backend/src/test/java/pl/cyfronet/s4e/data/repository/GroupRepositoryTest.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    private void reset() {
+        institutionRepository.deleteAll();
+        appUserRepository.deleteAll();
+    }
+
+    @Test
+    void shouldFindAllMembersEmails() {
+        val membersEmails = groupRepository.findAllMembersEmails(slugInstitution,"default", AppRole.GROUP_MEMBER);
```
Comment (jswk) [inline s4e-backend/src/test/java/pl/cyfronet/s4e/data/repository/GroupRepositoryTest.java]: Formatting: whitespace

File: s4e-backend/src/test/java/pl/cyfronet/s4e/data/repository/GroupRepositoryTest.java [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        assertThat(membersEmails.isEmpty(), is(false));
```
Comment (jswk) [inline s4e-backend/src/test/java/pl/cyfronet/s4e/data/repository/GroupRepositoryTest.java]: Unnecessary

File: s4e-backend/src/test/java/pl/cyfronet/s4e/data/repository/AppUserRepositoryTest.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+                .group(group)
+                .build());
+    }
+
+    @AfterEach
+    void tearDown() {
+        reset();
+    }
+
+    private void reset() {
```
Comment (jswk) [inline s4e-backend/src/test/java/pl/cyfronet/s4e/data/repository/AppUserRepositoryTest.java]: Use TestDBHelper

File: s4e-backend/src/test/java/pl/cyfronet/s4e/data/repository/GroupRepositoryTest.java [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+                .group(group)
+                .build());
+    }
+
+    @AfterEach
+    void tearDown() {
+        reset();
+    }
+
+    private void reset() {
```
Comment (jswk) [inline s4e-backend/src/test/java/pl/cyfronet/s4e/data/repository/GroupRepositoryTest.java]: Use TestDBHelper

File: s4e-backend/src/test/java/pl/cyfronet/s4e/service/GroupServiceTest.java [hunk 7]
```diff
@@ -215,20 +210,28 @@ public void shouldUpdateUsersGroupsIfUserInTwoInstitutionsAndRemoveOne() throws
...
         assertThat(groupService.getMembers("instytucja-22", "group-21"), hasSize(1));
 
-        groupService.updateUsersGroups(user.getId(), "instytucja-12", Set.of());
+        groupService.updateUserGroups(user.getId(), "instytucja-12", Set.of());
 
         assertThat(groupService.getMembers("instytucja-12", "group-13"), empty());
         assertThat(groupService.getMembers("instytucja-22", "group-21"), hasSize(1));
     }
+
+    private UserRole buildUserRoleObject(AppUser user, Group group){
```
Comment (jswk) [inline s4e-backend/src/test/java/pl/cyfronet/s4e/service/GroupServiceTest.java]: `buildGroupMemberUserRole`?

File: s4e-backend/src/main/java/pl/cyfronet/s4e/controller/AppUserController.java [hunk 8]
```diff
@@ -190,23 +171,23 @@
...
         }
         return ResponseEntity.ok().build();
     }
 
     @Operation(summary = "Get user profile")
     @ApiResponses({
             @ApiResponse(responseCode = "200", description = "User was retrieved"),
-            @ApiResponse(responseCode = "403", description = "Forbidden: Don't have permission to get profile")
+            @ApiResponse(responseCode = "403", description = "Forbidden: Don't have permission to get profile"),
+            @ApiResponse(responseCode = "404", description = "Forbidden: Don't have permission to get profile")
```
Comment (jswk) [inline s4e-backend/src/main/java/pl/cyfronet/s4e/controller/AppUserController.java]: Don't have permission? Rather user not found, however, it shouldn't happen. I would omit the 404 code here.

Comment (kmarszalek) [reply s4e-backend/src/main/java/pl/cyfronet/s4e/controller/AppUserController.java]: msg changed, we throw notfound, so i chose to let it be the way it was :)

**Pull request merged**